point.go

```diff
@@ -0,0 +1,306 @@
+		return buf, rem, surge.ErrUnexpectedEndOfBuffer
+	}
+
+	err := p.SetBytes(buf[:PointSizeMarshalled])
+	if err != nil {
+		return buf[PointSizeMarshalled:], rem - PointSize, err
+	}
+
+	return buf[PointSizeMarshalled:], rem - PointSize, nil
+}
+
+// IsInfinity returns true if the point represents the point at infinity, and
+// false otherwise.
+func (p *Point) IsInfinity() bool {
+	return p.inner.infinity != 0
+}
+
+// IsOnCurve returns true if the point is on the elliptic curve, and false
+// otherwise. The point at infinity will return false.
+func (p *Point) IsOnCurve() bool {
+	var tmp C.secp256k1_ge
+	C.secp256k1_ge_set_gej(&tmp, &p.inner)
+	return C.secp256k1_ge_is_valid_var(&tmp) != 0
+}
+
+// Eq returns true if the two curve points are equal, and false otherwise.
+func (p *Point) Eq(other *Point) bool {
+	if p.IsInfinity() != other.IsInfinity() {
+		return false
+	}
+
+	if p.IsInfinity() {
+		return true
+	}
+
+	// Scale p so that the z fields are equal. Once the z fields are equal, the
+	// points will be equal if and only if the x and y fields are equal.
+	//
+	// NOTE: This modifies the representation of p, but it will still represent
+	// the same point on the elliptic curve.
+	var s C.secp256k1_fe
+	C.secp256k1_fe_inv(&s, &p.inner.z)
+	C.secp256k1_fe_mul(&s, &s, &other.inner.z)
+	C.secp256k1_gej_rescale(&p.inner, &s)
+	normalizeXYZ(&p.inner)
+
+	return fpEq(&p.inner.x, &other.inner.x) && fpEq(&p.inner.y, &other.inner.y)
+}
+
+// BaseExp computes the scalar multiplication of the canonical generator of the
+// curve by the given scalar.
+func (p *Point) BaseExp(scalar *Fn) {
+	scalarMul(&p.inner, &C.secp256k1_generator, &scalar.inner)
+}
+
+// Scale computes the scalar multiplication of the given curve point by the
+// given scalar.
+//
+//NOTE: It is assumed that the input point is not the point at infinity.
+func (p *Point) Scale(a *Point, scalar *Fn) {
+	var tmp C.secp256k1_ge
+	C.secp256k1_ge_set_gej(&tmp, &a.inner)
+	scalarMul(&p.inner, &tmp, &scalar.inner)
+}
+
+// ScaleExt is the same as Scale but also works when the input point represents
+// the point at infinity; in this case the result of the scalar multiplication
+// will also be the point at infinity.
+func (p *Point) ScaleExt(a *Point, scalar *Fn) {
+	if a.IsInfinity() {
+		p.inner = a.inner
+		return
+	}
+
+	p.Scale(a, scalar)
+}
+
+func scalarMul(dst *C.secp256k1_gej, a *C.secp256k1_ge, scalar *C.secp256k1_scalar) {
+	// The final argument should be the maximum bit length of the absolute
+	// value of the scalar plus one, hence 256 + 1.
+	C.secp256k1_ecmult_const(dst, a, scalar, 257)
+
+	// The curve scalar multiplication function doesn't make sure that the
+	// coordinates are normalized, so we need to do this manually.
+	normalizeXYZ(dst)
+}
+
+// Add computes the curve addition of the two given curve points.
+func (p *Point) Add(a, b *Point) {
+	C.secp256k1_gej_add_var(&p.inner, &a.inner, &b.inner, C.null_ptr)
```

Comment:
Will this C function panic correctly if `a` or `b` is `nil`?

Reply:
Same question for the other functions accepting pointers.

Reply:
Good point, forgot about the pattern that was used for `Fn` and `Fp`, will fix this.

Reply:
Fixed in 893f83c.